### ISSUE TITLE
'hasp.py' unable to parse #EXT-X-CUE-OUT, #EXT-X-CUE-OUT-CONT, #EXT-X-CUE-IN

### ISSUE THREAD

**Issue by moving-digital** [futzu/threefive#51]
Hi @futzu 

Following up with another issue, `hasp.py` is unable to parse the SCTE-35 HLS cues in my two m3u8 playlists:
- #EXT-X-CUE-OUT
- #EXT-X-CUE-OUT-CONT
- #EXT-X-CUE-IN

Playlist 1 has the following syntax:
```
#EXT-X-CUE-OUT:90.023267
#EXT-X-CUE-OUT-CONT:6.006/90.023267
(..... Removed on purpose ....)
#EXT-X-CUE-OUT-CONT:78.078/90.023267
#EXTINF:6.006,
#EXT-X-CUE-OUT-CONT:84.084/90.023267
#EXTINF:5.939267,
#EXT-X-CUE-IN
#EXTINF:4.004,
#EXTINF:6.006,
```

While Playlist 2 has a different structure:

```
#EXTINF:4.0,
sftv_480p/media-ucvpbplas_228029.ts
#EXTINF:1.12,
sftv_480p/media-ucvpbplas_228030.ts
#EXT-OATCLS-SCTE35:c3BsaWNlLTY3MTE0OTQw
#EXT-X-ASSET:CAID=splice-67114940
#EXT-X-CUE-OUT:30.00
#EXT-X-DATERANGE:ID="splice-67114940",START-DATE="2022-02-22T04:33:38.99782700Z",PLANNED-DURATION=30.00,SCTE35-OUT=0x000000000000000000000000000000323032322D30322D32325430343A33333A33382E39393738323730305A
#EXTINF:7.0,
sftv_480p/media-ucvpbplas_228031.ts
#EXT-X-CUE-OUT-CONT:ElapsedTime=7.00,Duration=30.00,SCTE35=c3BsaWNlLTY3MTE0OTQw
#EXTINF:4.0,
sftv_480p/media-ucvpbplas_228032.ts
#EXT-X-CUE-OUT-CONT:ElapsedTime=11.00,Duration=30.00,SCTE35=c3BsaWNlLTY3MTE0OTQw
#EXTINF:4.0,
sftv_480p/media-ucvpbplas_228033.ts
#EXT-X-CUE-OUT-CONT:ElapsedTime=15.00,Duration=30.00,SCTE35=c3BsaWNlLTY3MTE0OTQw
#EXTINF:4.0,
sftv_480p/media-ucvpbplas_228034.ts
#EXT-X-CUE-OUT-CONT:ElapsedTime=19.00,Duration=30.00,SCTE35=c3BsaWNlLTY3MTE0OTQw
#EXTINF:4.0,
sftv_480p/media-ucvpbplas_228035.ts
#EXT-X-CUE-OUT-CONT:ElapsedTime=23.00,Duration=30.00,SCTE35=c3BsaWNlLTY3MTE0OTQw
#EXTINF:4.0,
sftv_480p/media-ucvpbplas_228036.ts
#EXTINF:4.0,
sftv_480p/media-ucvpbplas_228036.ts
#EXT-X-CUE-OUT-CONT:ElapsedTime=27.00,Duration=30.00,SCTE35=c3BsaWNlLTY3MTE0OTQw
#EXTINF:3.0,
sftv_480p/media-ucvpbplas_228037.ts
#EXT-X-CUE-IN
#EXT-X-DATERANGE:ID="splice-67114941",DURATION=0.00,SCTE35-IN=0x000000000000000000000000000000323032322D30322D32325430343A33343A30382E39393738323730305A
#EXTINF:5.0,
sftv_480p/media-ucvpbplas_228038.ts
```

None of them can be parsed with `hasp.py`, could you have a look? I can provide you the m3u8 files.

**Comment by futzu** [comment]
```
SCTE35=c3BsaWNlLTY3MTE0OTQw
```
I can look at that and tell you it's not SCTE-35. 
```
IN=0x000000000000000000000000000000323032322D30322D32325430343A33343A30382E39393738323730305A
```
that looks kind of  kooky as well/

What made that file?

**Comment by moving-digital** [comment]
> ```
> SCTE35=c3BsaWNlLTY3MTE0OTQw
> ```
> 
> I can look at that and tell you it's not SCTE-35.
> 
> ```
> IN=0x000000000000000000000000000000323032322D30322D32325430343A33343A30382E39393738323730305A
> ```
> 
> that looks kind of kooky as well/
> 
> What made that file?

Hi @futzu  I wish I could tell which packager, but I have no idea. Yet this stream is currently live and working well...

Are you able to discard the unnecessary parameters for hasp.py not to crash?

**Comment by moving-digital** [comment]
Hi @futzu 

**For playlist 1** (here: [karaoke_scte35.m3u8.txt](https://github.com/futzu/scte35-threefive/files/8124005/karaoke_scte35.m3u8.txt) :

No SCTE-35 can be parsed even if all the markers are valid. Could you please check?

**For playlist 2** (here: [ftv_scte35.m3u8.txt](https://github.com/futzu/scte35-threefive/files/8124001/ftv_scte35.m3u8.txt) :

I was having a look at the [hlsparse.py](https://github.com/futzu/scte35-threefive/blob/master/examples/hls/hlsparse.py) file and I also found a reference to the same long hexadecimal value, line 41:

`##EXT-X-DATERANGE:ID="splice-6FFFFFF0",START-DATE="2014-03-05T11:15:00Z",PLANNED-DURATION=59.993,SCTE35-OUT=0xFC002F0000000000FF000014056FFFFFF000E011622DCAFF000052636200000000000A000829896F50000008700000000`

So I guess this is a valid one. I edited my own m3u8 with the above hexa value, but it also can't parse any SCTE-35 cues and it obviously still throws the same error:

```
Traceback (most recent call last):
  File "hasp.py", line 236, in <module>
    HASP(arg).decode()
  File "hasp.py", line 221, in decode
    self.do_segment(line)
  File "hasp.py", line 202, in do_segment
    self.show_segment_times(stanza)
  File "hasp.py", line 187, in show_segment_times
    stanza.do_cue()
  File "hasp.py", line 123, in do_cue
    tf = threefive.Cue(self.cue)
  File "/usr/local/lib/pypy3.7/dist-packages/threefive/cue.py", line 56, in __init__
    self.bites = self._mk_bits(data)
  File "/usr/local/lib/pypy3.7/dist-packages/threefive/cue.py", line 257, in _mk_bits
    bites = int.to_bytes(i, i_len, byteorder="big")
OverflowError: int too big to convert
```

However I agree with you that the `#EXT-OATCLS-SCTE35:c3BsaWNlLTY3MTE0OTQw` seems weird... If it is not needed for the SCTE-35 cue parsing, could we just discard it ?

**Comment by futzu** [comment]
I never got your files, the links don't work. 

Dude you need to read the SCTE-35 and hls specifications.

**Comment by futzu** [comment]
```
 IN=0x000000000000000000000000000000323032322D30322D32325430343A33343A30382E39393738323730305A
```
Is not SCTE-35, I know that because there is no table ID at the start, or splice info section. 
It is also not the  'Too big' int. 


```
>>>> IN
1353072972795142803346103647730042803828023333179364518029600318566490

```
You haven't given me enough information to even diagnose the problem. 
**If want my help, I need the m3u8 and the stream.**  
If you trust me enough to use threefive, then you should
trust me enough to show me the information I need.

**Comment by futzu** [comment]
just email me spam@futzu.com

**Comment by moving-digital** [comment]
> just email me [spam@futzu.com](mailto:spam@futzu.com)

@futzu  Email sent!

**Comment by futzu** [comment]
![ext-x-daterange](https://user-images.githubusercontent.com/52701496/155646796-fcd286e9-9ea3-4c0e-9d5d-99e182ecb898.png)
![scte35-hls-cue-tags-one](https://user-images.githubusercontent.com/52701496/155646828-a11e07ce-0897-4498-aa51-ac0faf3b3563.png)
![scte35-hls-cue-tags-two](https://user-images.githubusercontent.com/52701496/155646877-9fba8d0c-d5c3-4192-83d2-c1670bc1d848.png)

**Comment by moving-digital** [comment]
Thanks @futzu !

Understood for playlist 2.

But how about playlist 1 ?

**Comment by moving-digital** [comment]
Hi @futzu 

Would you happen to have an update about the issue with the following playlist ?

 karaoke_scte35.m3u8.txt

**Comment by futzu** [comment]
Threefive is working correctly.
It's not parsing it because it's not actually scte-35.

If you want me to sort it out for you, I do consulting. Send me an email and we can discuss terms.


-------- Original Message --------

**Comment by futzu** [comment]
* The karaoke one doesnt have a SCTE-35 Cue, 
* #EXT-CUE-OUT-CONT is incomplete and inaccurate.

I can go on and on, but I'm not going to do your job for you, unless you pay me.

**Comment by futzu** [comment]
Read this.
https://datatracker.ietf.org/doc/html/rfc8216#section-4.3.2.7.1

**Comment by moving-digital** [comment]
Hi @futzu 

I actually agree with you that this kind of SCTE-35 is not as per the specs, but I must say this is the sad truth about SCTE-35 in HLS stream.

As per [bbgdzxng1](https://github.com/bbgdzxng1), I quote :

> There are many different syntax formats for manifest decoration. It is a mess.
> 
> There are only two "standard" ones, #EXT-X-DATERANGE (as per RFC8216 bis) and #EXT-X-SCTE35 (ANSI/SCTE 35 2019a). However, support for these by segmenting encoder vendors and ad insertion vendors has historically been very low, although support for #EXT-X-DATERANGE is now growing.
> 
> Then there are the vendor implementations. Each detecting vendor does their own thing, but both Google Ad Manager and MediaTailor do a reasonable of documenting the formats they can detect:
> https://docs.aws.amazon.com/mediatailor/latest/ug/hls-ad-markers.html
> https://docs.aws.amazon.com/medialive/latest/ug/sample-manifests-hls.html
> https://support.google.com/admanager/answer/7245661?hl=en#supported-ad-markers
> 
> - #EXT-X-CUE (Adobe Primetime)
> - #EXT-X-CUE-OUT, & #EXT-X-CUE-OUT-CONT with two sub variations
> - #EXT-X-CUE-OUT:DURATION=10.50
> - #EXT-X-CUE-OUT:10.50
> - #EXT-OATCLS-SCTE35 (aka Enhanced, same as Elemental but with additional binary payload)
> - #EXT-X-SPLICEPOINT-SCTE35
> And there are several other proprietary segmenting encoder vendor implementations as well.
> Encoder support for #EXT-X-DATERANGE (as per RFC8216) is on the rise, because it is RFC. And it is the real SCTE message.
> 
> #EXT-X-CUE-OUT/IN is the simplest markup format, since an encoder/segmenter can implement this without even reading the SCTE-35 documentation. This is the most common "simple markup".
> 
> The danger point for the implementer is that a manifest can declare the duration in multiple tag events. The duration is often included in the OUT message, the duration is included as a rolling update in the CONT message, the duration is included in the binary payload. The binary payload may then be repeated/updated between tags. A client such as TS Duck needs to choose one as the point of authority, decide some level of precedence if the same contradictory info is included in the same tag (yes it happens), and decide whether "last known" wins. And log them all so that it is debuggable.
> 
> To complicate it further, any declared duration should be trumped by a CUE-IN, because that is conceptually a splice immediate return to network. But the duration could also expire before the CUE-IN, especially when using non-integer framerates like 30000/1001. So in that case, the declared duration should win over the CUE-IN.
> 
> Worse than that, a segmenting encoder may choose to send more than one manifest decoration tag format. Because more is better, right?
> 
> So which one could a tool choose, support and even implement?
> 
> - Allow user to control whether the SCTE-35 data track or HLS manifest is considered the reference
> - If HLS manifest, which markup formats/tags should be supported?
> - When contradictory information is received, what takes precedence? And log it.
> - How do you debug contradictory info?
> - Log unsupported tags events, so it is clear that unsupported tags are ignored.
>
> Supporting all the various vendor formats would be a Sisyphean task. 

To make threefive robust and future-proof, that may be worth looking into implementing all different ways to handle SCTE-35 in HLS, otherwise its field of action for HLS may be limited.

**Comment by futzu** [comment]
I am not trying to be a dick, but you don't understand how all this works. 

Non-integer frame rates are what is most often used. 

You are confusing HLS Segment duration with SCTE-35 Splice duration, two totally different things. 

what you call "binary payload" is what I call a SCTE-35 cue. 

**All of the HLS SCTE-35 tags you mentioned above require SCTE-35 cues,**

**Your HLS tags either do not have SCTE-35 cues (aka binary payload) at all, 
or they have incorrect SCTE-35 cues that cannot be parsed.**

threefive will never support proprietary implementations, 
however, I am asked to write proprietary parsers often.

You can spend months figuring it out, or you can hire me to do it for you in a week or two.

**Comment by futzu** [comment]
I did find something I was doing wrong with your HLS. 
The segment uri starts with ../../../../, and the hasp.py script was getting 500 errors back from the server when it asked for segments.

like this

 ```
https://example.com/a/b/c/d/../../../../segment.ts
````

I updated hasp.py last night so that it handles that before it makes the request

like this 

```
https://examples.com/segment.ts
```

Heres the [code](https://github.com/futzu/scte35-threefive/blob/master/examples/hls/hasp.py#L52).
